feat: Add Unsplash Background Support (#239)

Satorino01 opened the pull request:
### Overview
This PR introduces an option to fetch random background images from Unsplash, allowing users to customize the quote's backdrop for a more visually appealing design.

### Changes
- Implemented a new feature that uses `bgSource=unsplash` and an optional `unsplashQuery` parameter to retrieve background images.
- Updated README.md with usage instructions and a sample code snippet for Unsplash backgrounds.
- If `bgSource=unsplash` is detected, the default background color becomes transparent.

### How to Test

1. Add your [Unsplash API](https://unsplash.com/developers) Access Key to the `.env` file as `UNSPLASH_ACCESS_KEY`.
2. Run `npm start` to launch the server.
3. Navigate to `http://localhost:3004/quote?bgSource=unsplash&unsplashQuery=nature.`

**Below is an example screenshot showing the feature in action:**
![スクリーンショット 2025-02-17 3 15 02](https://github.com/user-attachments/assets/7fb9f18b-3a3b-4949-98fe-ab3152b3eae7)

### Additional Considerations

- There is a readability issue when the chosen background image has colors similar to the text color. Because the final background is unknown beforehand, we might need a semi-transparent overlay behind the text or add a text outline/drop-shadow to improve legibility. Suggestions or feedback on how to address this are welcome!

- Currently, when bgSource=unsplash is set, bgColor is automatically overridden to transparent. However, if the Unsplash API fails or returns a "Rate Limit Exceeded" error, the background image may not load. Should we maintain a non-transparent fallback color in such cases instead of forcing transparency? Feedback on this approach would be appreciated.

### Sample Images 

```
http://localhost:3004/quote?bgSource=unsplash&unsplashQuery=morning&fontColor=%23000000
```
<img width="768" alt="スクリーンショット 2025-02-17 10 37 45" src="https://github.com/user-attachments/assets/b319ee06-42e0-4323-b6de-f0b3a00ac215" />

-----
```
http://localhost:3004/quote?bgSource=unsplash&unsplashQuery=tech&fontColor=%23FFFFFF
```
<img width="768" alt="スクリーンショット 2025-02-17 10 34 15" src="https://github.com/user-attachments/assets/d2c9d47e-6469-4ec4-9c30-0cb9755a3aa5" />

-----
```
http://localhost:3004/quote?bgSource=unsplash&unsplashQuery=night&fontColor=%23FFFFFF
```
<img width="770" alt="スクリーンショット 2025-02-17 10 26 33" src="https://github.com/user-attachments/assets/edbd2125-c4a9-44cf-9230-5ea9492be9d8" />

-----
```
http://localhost:3004/quote?bgSource=unsplash&unsplashQuery=sea&fontColor=%23FFFFFF
```
<img width="768" alt="スクリーンショット 2025-02-17 10 31 14" src="https://github.com/user-attachments/assets/78163a68-829d-479c-b696-ede8f61489fc" />


Closes #239

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
  - Introduced Unsplash background support, enabling you to enhance quote images with dynamic visuals. Now, by specifying the new parameters, you can request random or keyword-based backgrounds from Unsplash, adding a fresh look to your quotes.

- **Documentation**
  - Updated documentation with a dedicated section on Unsplash backgrounds, providing clear instructions and example snippets to help you customize your quotes effortlessly.

- **Improvements**
  - Enhanced background handling logic to support dynamic background colors based on the source specified.
  - Added functionality to sanitize background image URLs for improved security and reliability.
  - Introduced new state management in the Home component to allow user selection of background source and query.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Satorino01 commented on src/api/services/quotesService.js:
```diff
@@ -62,3 +68,20 @@ const getQuote = async (quoteObj) => {
   getQuote,
 };
+
+async function getUnsplashImage(query) {
+  const accessKey = process.env.UNSPLASH_ACCESS_KEY;
+  const unsplashUrl = `https://api.unsplash.com/photos/random?query=${encodeURIComponent(query)}&client_id=${accessKey}`;
+
+  try {
+    const response = await requestApi(unsplashUrl);
+    if (response && response.urls && response.urls.regular) {
+      return response.urls.regular;
+    } else {
+      return '';
+    }
+  } catch (err) {
+    console.error('Error fetching Unsplash image:', err);
+    return '';
+  }
+}
```
The code already fails at [fetchApi.js#L7](https://github.com/shravan20/github-readme-quotes/blob/094738e7ead50e0e96f7799de6c2b54bc1bcd11c/src/utils/fetchApi.js#L7) when Unsplash returns a “Rate Limit Exceeded” response, so it never reaches the [getUnsplashImage() catch block](https://github.com/shravan20/github-readme-quotes/blob/094738e7ead50e0e96f7799de6c2b54bc1bcd11c/src/api/services/quotesService.js#L77). Splitting out the 429 status check here has no effect.

Satorino01 commented on src/api/controllers/quotesController.js:
```diff
@@ -14,7 +14,10 @@ const quoteController = async (req, res, next) => {
     }
     const bgColor = req.query.bgColor;
-    if (bgColor) {
+    const bgSource = req.query.bgSource || '';
+    if (bgSource === 'unsplash') {
+      theme.bg_color = 'transparent';
+    } else if (bgColor) {
       theme.bg_color = bgColor;
     }
```
1. We’ll let the reviewer decide whether a fallbackBgColor is necessary. If they deem it required, we’ll implement it.

2. We’ll also let the reviewer decide on the 100-character query limit, since Unsplash’s documentation doesn’t mention any specific length recommendation.
https://help.unsplash.com/en/?q=query+length

Satorino01 commented on .env.example:
```diff
@@ -1,4 +1,5 @@
 SERVICE_URL = <URL_FETCHING_RANDOM_QUOTES>
-
+UNSPLASH_ACCESS_KEY = <YOUR_UNSPLASH_ACCESS_KEY>
+UNSPLASH_SECRET_KEY = <YOUR_UNSPLASH_SECRET_KEY>
```
I addressed that in commit 094738e.

shravan20 commented:
Greetings @Satorino01 

Thank you so much for your contribution. Appreciate your efforts and putting up time for the project. ♥️👏🏼 


I will have this reviewed today. In the mean time, can you share some sample images of the current changes as reference for other fellow reviewers too.

Thank you again!!

Satorino01 commented:
Hello @shravan20 

Thank you for the quick reply. I’ve added some extra screenshots in the PR description showing different queries and layouts with Unsplash backgrounds. Please let me know if you need more examples or further details!

shravan20 commented:
@Satorino01: Overall LGTM. Amazing work!! Really loved it!! <3 ❇️ 

Can we do two things here:
1. To avoid breaching rate limit of the unsplash API, can we try to cache the image at application level, for 10-15 mins for given query?
2. Can we include this change as part of the UI as well so that folks can configure and preview these changes via react UI too?

Satorino01 commented:
Hello @shravan20 

Thank you for the feedback and kind words! I’d be happy to incorporate both suggestions: implementing a 10–15 minute cache for repeated queries and adding a React UI option so users can configure and preview these features. I’ll start working on them this weekend and will let you know once I have something ready.

Satorino01 commented on src/api/controllers/quotesController.js:
```diff
@@ -33,6 +36,7 @@ const quoteController = async (req, res, next) => {
     let font = fonts[req.query.font] ? fonts[req.query.font] : fonts['default'];
 
     let quoteType = req.query.quoteType || '';
+    let unsplashQuery = req.query.unsplashQuery || '';
```
NOTE: If multiple unsplashQuery params are given 
(e.g. ?unsplashQuery=morning&unsplashQuery=night),
this may arrive as an array instead of a single string.

Satorino01 commented:
@shravan20
I’ve just pushed two commits addressing your suggestions
- 1369bb5 : Implements the 10–15 minute query cache to prevent excessive Unsplash requests.
- 72548d6 : Adds a new React UI option, letting users configure and preview Unsplash backgrounds.

I’d appreciate it if you could review these changes whenever you have time. Let me know if there’s anything else you’d like me to adjust!

shravan20 commented:
@Satorino01: Thank you so much for your contribution!!
Love your work!!
